File: app/src/main/java/io/github/mohamedisoliman/androiddev/ui/post/PostActivity.java
package io.github.mohamedisoliman.androiddev.ui.post;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import io.github.mohamedisoliman.androiddev.R;
import io.github.mohamedisoliman.androiddev.data.model.RedditPost;
import io.github.mohamedisoliman.androiddev.databinding.ActivityPostBinding;
import io.github.mohamedisoliman.androiddev.ui.Util;

/**
 * Created by Mohamed Ibrahim on 8/5/18.
 */
public class PostActivity extends AppCompatActivity {

  private static final String KEY_POST = "POST";
  ActivityPostBinding binding;
  private RedditPost post;

  public static void start(Context context, RedditPost post) {
    Intent starter = new Intent(context, PostActivity.class);
    starter.putExtra(KEY_POST, post);
    context.startActivity(starter);
  }

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    post = getIntent().getParcelableExtra(KEY_POST);
    binding = DataBindingUtil.setContentView(this, R.layout.activity_post);
    setSupportActionBar(binding.appbar.toolbar);
    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    getSupportActionBar().setDisplayShowTitleEnabled(false);

    binding.webview.setWebViewClient(getWebViewClient());
    binding.webview.setWebChromeClient(getWebChromeClient());
    binding.webview.loadUrl(post.getUrl());
  }

  @NonNull
  private WebChromeClient getWebChromeClient() {
    return new WebChromeClient() {

      @Override
      public void onProgressChanged(WebView view, int newProgress) {
        setValue(newProgress);
        super.onProgressChanged(view, newProgress);
      }
    };
  }

  @NonNull
  private WebViewClient getWebViewClient() {
    return new WebViewClient() {
      @Override
      public boolean shouldOverrideKeyEvent(WebView view, KeyEvent event) {
        return true;
      }
    };
  }

  public void setValue(int newProgress) {
    binding.appbar.progress.setProgress(newProgress);
    if (newProgress == 100) {
      binding.appbar.progress.setVisibility(View.GONE);
    } else if (binding.appbar.progress.getVisibility() == View.GONE) {
      binding.appbar.progress.setVisibility(View.VISIBLE);
    }
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.menu_post_activity, menu);
    return super.onCreateOptionsMenu(menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case android.R.id.home:
        finish();
        break;
      case R.id.share:
        if (post != null) {
          Util.sharePost(post, this);
        }
        break;
      case R.id.external:
        if (post != null) {
          Util.openExternal(post.getUrl(), this);
        }
        break;
    }
    return super.onOptionsItemSelected(item);
  }
}

File: app/src/main/java/io/github/mohamedisoliman/androiddev/ui/post/PostActivity.kt
package io.github.mohamedisoliman.androiddev.ui.post

import android.content.Context
import android.content.Intent
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import io.github.mohamedisoliman.androiddev.R
import io.github.mohamedisoliman.androiddev.data.model.RedditPost
import io.github.mohamedisoliman.androiddev.databinding.ActivityPostBinding
import io.github.mohamedisoliman.androiddev.ui.openExternal
import io.github.mohamedisoliman.androiddev.ui.sharePost

/**
 * Created by Mohamed Ibrahim on 8/5/18.
 */
class PostActivity : AppCompatActivity() {
  private lateinit var binding: ActivityPostBinding
  private lateinit var post: RedditPost

  private val webChromeClient: WebChromeClient
    get() = object : WebChromeClient() {

      override fun onProgressChanged(view: WebView, newProgress: Int) {
        setValue(newProgress)
        super.onProgressChanged(view, newProgress)
      }
    }

  private val webViewClient: WebViewClient
    get() = object : WebViewClient() {
      override fun shouldOverrideKeyEvent(view: WebView, event: KeyEvent): Boolean {
        return true
      }
    }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    post = intent.getParcelableExtra(KEY_POST)
    binding = DataBindingUtil.setContentView(this, R.layout.activity_post)
    setSupportActionBar(binding.appbar?.toolbar)
    supportActionBar?.setDisplayHomeAsUpEnabled(true)
    supportActionBar?.setDisplayShowTitleEnabled(false)

    binding.webview.webViewClient = webViewClient
    binding.webview.webChromeClient = webChromeClient
    binding.webview.loadUrl(post.url)
  }

  fun setValue(newProgress: Int) {
    binding.appbar?.progress?.progress = newProgress
    binding.appbar?.progress?.visibility = View.VISIBLE
    when (newProgress) {
      100 -> binding.appbar?.progress?.visibility = View.GONE
      else -> binding.appbar?.progress?.visibility = View.VISIBLE

    }
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu_post_activity, menu)
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      android.R.id.home -> finish()
      R.id.share -> sharePost(post)
      R.id.external -> openExternal(post.url!!)
    }
    return super.onOptionsItemSelected(item)
  }

  companion object {

    private const val KEY_POST = "POST"

    fun start(context: Context, post: RedditPost) {
      val starter = Intent(context, PostActivity::class.java)
      starter.putExtra(KEY_POST, post)
      context.startActivity(starter)
    }
  }
}
